[WIP] Add JWT authentication with user creation

DrissiReda opened the pull request:
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[![Open in GitHub Codespaces](https://github.com/codespaces/badge.svg)](https://codespaces.new/DrissiReda/mlflow/pull/14144?quickstart=1)

#### Install mlflow from this PR

```
# Use `%sh` to run this command on Databricks
OPTIONS=$(if pip freeze | grep -q 'mlflow @ git+https://github.com/mlflow/mlflow.git'; then echo '--force-reinstall --no-deps'; fi)
pip install $OPTIONS git+https://github.com/mlflow/mlflow.git@refs/pull/14144/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 14144
```

</p>
</details>

### Related Issues/PRs

<!-- Resolve --> 

### What changes are proposed in this pull request?

I want to add the ability to log in via JWT and allow for user creation if they don't already exist, 
if maintainers are interested I can even add role -> permissions mapping to bring mlflow one step closer to OIDC support.

This is still a work in progress and any help is welcome especially to be able to interact with the authentication database
without HTTP requests.

### How is this PR tested?

- [ ] Existing unit/integration tests
- [ ] New unit/integration tests
- [x] Manual tests

Currently test code is still in PR but it'll be moved to a proper test if it gets traction

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [x] Yes. I'll have to add doc related to configuration:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

### Release Notes

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

JWT authentication with automatic local user creation.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [x] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [x] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

#### Should this PR be included in the next patch release?

`Yes` should be selected for bug fixes, documentation updates, and other small changes. `No` should be selected for new features and larger changes. If you're unsure about the release classification of this PR, leave this unchecked to let the maintainers decide.

<details>
<summary>What is a minor/patch release?</summary>

- Minor release: a release that increments the second part of the version number (e.g., 1.2.0 -> 1.3.0).
  Bug fixes, doc updates and new features usually go into minor releases.
- Patch release: a release that increments the third part of the version number (e.g., 1.2.0 -> 1.2.1).
  Bug fixes and doc updates usually go into patch releases.

</details>

<!-- patch -->

- [ ] Yes (this PR will be cherry-picked and included in the next patch release)
- [x] No (this PR will be included in the next minor release)

Commit:
Add jwt authentication with user creation

ion-elgreco commented:
@DrissiReda fyi, there is an oss oidc plugin, https://github.com/data-platform-hq/mlflow-oidc-auth

DrissiReda commented:
> @DrissiReda fyi, there is an oss oidc plugin, https://github.com/data-platform-hq/mlflow-oidc-auth

Thanks I checked it out but it seems broken from the issues, and is very limited in regards to features, I'm more interested with synchronizing users 

Can you give me a hint on how I can create users internally in my module without having to run an http request?

ion-elgreco commented:
> > @DrissiReda fyi, there is an oss oidc plugin, https://github.com/data-platform-hq/mlflow-oidc-auth
> 
> Thanks I checked it out but it seems broken from the issues, and is very limited in regards to features, I'm more interested with synchronizing users 
> 
> Can you give me a hint on how I can create users internally in my module without having to run an http request?

I have gotten it working eventually, was a bit of a hassle because the package doesn't throw useful errors atm, so I had to dig in the code

DrissiReda commented:
> > > @DrissiReda fyi, there is an oss oidc plugin, https://github.com/data-platform-hq/mlflow-oidc-auth
> > 
> > 
> > Thanks I checked it out but it seems broken from the issues, and is very limited in regards to features, I'm more interested with synchronizing users
> > Can you give me a hint on how I can create users internally in my module without having to run an http request?
> 
> I have gotten it working eventually, was a bit of a hassle because the package doesn't throw useful errors atm, so I had to dig in the code

thanks, but since I'll be doing some coding eventually I'd rather do that here.

DrissiReda committed:
Improve logging, add syncing perms to roles, change local user's password.

- Check the token's validity.
- remove testing hardcoded tokens
- Improve logging
- Sync permissions of every object to user roles
- Use keycloak or oidc provider's unique user id as basic auth password. This means any mechanisms with a token can log in via basic auth as well.

Signed-off-by: Reda DRISSI <reda.drissi@yahoo.fr>
